netlogo-gui/src/main/app/WidgetActions.scala

```diff
@@ -0,0 +1,166 @@
+      bounds.y -= WidgetWrapper.BORDER_N
+      bounds.height += WidgetWrapper.BORDER_N + WidgetWrapper.BORDER_S
+      bounds
+    })(collection.breakOut)
+    val widgets = widgetPanel.widgetsBeingDragged
+    widgetPanel.dropSelectedWidgets()
+    val finalMap: Map[WidgetWrapper, Rectangle] = widgets.map(a => a -> a.getBounds())(collection.breakOut)
+    undoManager.addEdit(new MoveWidgets(widgetPanel, widgets, initialMap, finalMap))
+  }
+
+  def resizeWidget(widgetWrapper: WidgetWrapper): Unit = {
+    val initialBounds = widgetWrapper.originalBounds
+    initialBounds.x -= WidgetWrapper.BORDER_E
+    initialBounds.width += WidgetWrapper.BORDER_E + WidgetWrapper.BORDER_W
+    initialBounds.y -= WidgetWrapper.BORDER_N
+    initialBounds.height += WidgetWrapper.BORDER_N + WidgetWrapper.BORDER_S
+
+    widgetWrapper.doDrop()
+    undoManager.addEdit(new ResizeWidget(widgetWrapper, initialBounds, widgetWrapper.getBounds()))
+  }
+
+  class AddWidget(widgetPanel: WidgetPanel, widgetWrapper: WidgetWrapper) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.reAddWidget(widgetWrapper)
+    }
+    override def undo(): Unit = {
+      widgetPanel.deleteWidget(widgetWrapper)
+    }
+  }
+
+  class RemoveWidget(widgetPanel: WidgetPanel, ww: WidgetWrapper) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.deleteWidget(ww)
+    }
+    override def undo(): Unit = {
+      widgetPanel.reAddWidget(ww)
+    }
+  }
+
+  class RemoveMultipleWidgets(widgetPanel: WidgetPanel, wws: Seq[WidgetWrapper]) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.deleteWidgets(wws)
+    }
+    override def undo(): Unit = {
+      for(ww <- wws){
+        widgetPanel.reAddWidget(ww)
+      }
+    }
+  }
+
+  class MoveWidgets(widgetPanel: WidgetPanel, wws: Seq[WidgetWrapper], initialMap: Map[WidgetWrapper, Rectangle], finalMap: Map[WidgetWrapper, Rectangle]) extends AbstractUndoableEdit {
+    override def redo:Unit = {
+      for(widgetWrapper <- wws){
+        val finalBound = new Rectangle(finalMap(widgetWrapper))
+        if(!widgetWrapper.selected()){
+          finalBound.x += WidgetWrapper.BORDER_E
+          finalBound.width -= WidgetWrapper.BORDER_E + WidgetWrapper.BORDER_W
+          finalBound.y += WidgetWrapper.BORDER_N
+          finalBound.height -= WidgetWrapper.BORDER_N + WidgetWrapper.BORDER_S
+        }
+        widgetWrapper.setBounds(finalBound)
+      }
+    }
+    override def undo(): Unit = {
+      for(widgetWrapper <- wws){
+        val initialBound = new Rectangle(initialMap(widgetWrapper))
+        if(!widgetWrapper.selected()) {
+          initialBound.x += WidgetWrapper.BORDER_E
+          initialBound.width -= WidgetWrapper.BORDER_E + WidgetWrapper.BORDER_W
+          initialBound.y += WidgetWrapper.BORDER_N
+          initialBound.height -= WidgetWrapper.BORDER_N + WidgetWrapper.BORDER_S
```

Comment:
These four lines show up 6 times in this file (or if they aren't the same four lines, they're very similar). A helper method would simplify the code a lot.